physionet-django/project/authorization/access.py

```diff
@@ -0,0 +1,100 @@
+from datetime import datetime
+
+from django.db.models import Q
+
+from events.models import Event, EventDataset
+from project.models import AccessPolicy, DUASignature, DataAccessRequest, PublishedProject
+from user.models import Training, TrainingType
+
+
+def get_accessible_projects(user):
+    """
+    Returns query filter for published projects accessible by a specified user
+    """
+    query = Q(access_policy=AccessPolicy.OPEN) & Q(deprecated_files=False)
+
+    dua_signatures = DUASignature.objects.filter(user=user)
+
+    if user.is_authenticated:
+        query |= Q(access_policy=AccessPolicy.RESTRICTED) & Q(
+            duasignature__in=dua_signatures
+        )
+
+    if user.is_credentialed:
+        completed_training = (
+            Training.objects.get_valid()
+            .filter(user=user)
+            .values_list("training_type")
+        )
+        not_completed_training = TrainingType.objects.exclude(
+            pk__in=completed_training
+        )
+        required_training_complete = ~Q(
+            required_trainings__in=not_completed_training
+        )
+
+        accepted_data_access_requests = DataAccessRequest.objects.filter(
+            requester=user, status=DataAccessRequest.ACCEPT_REQUEST_VALUE
+        )
+        contributor_review_with_access = Q(
+            access_policy=AccessPolicy.CONTRIBUTOR_REVIEW
+        ) & Q(data_access_requests__in=accepted_data_access_requests)
+
+        credentialed_with_dua_signed = Q(
+            access_policy=AccessPolicy.CREDENTIALED
+        ) & Q(duasignature__in=dua_signatures)
+
+        query |= required_training_complete & (
+            contributor_review_with_access | credentialed_with_dua_signed
+        )
+
+    # add projects that are accessible through events
+    events_all = Event.objects.filter(Q(host=user) | Q(participants__user=user))
```

Comment:
It feels messy for events to appear here, but perhaps the refactoring is just making the problems in our current approach more apparent (and so its a good thing that this seems awkward, because it is a step towards us fixing it!).